### PR TITLE
fix missing nvml in c10/cuda/driver_api.cpp issue

### PR DESCRIPTION
**Pull request by Fuzzkatt**
Since https://github.com/pytorch/pytorch/pull/99699 introduced a dependency on nvml for oom reporting in `c10/cuda/driver_api.h`, `c10/cuda/driver_api.cpp`, and `reportProcessMemoryInfo` from `c10/cuda/CUDACachingAllocator.cpp`, we've seen failures regarding cuda expandable segments and oom reporting in NVIDIA's internal CI, specifically on Jetson devices which don't have nvml support as it is incompatible with Jetson. Example failures using the latest upstream on Orin AGX node:

`python test/test_cuda.py -k test_notifies_oom` generates

```
Traceback (most recent call last):
  File "/usr/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.10/threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/pytorch/pytorch/test/test_cuda.py", line 1643, in _worker
    results[t] = torch.nn.functional.conv2d(results[t], weight, padding=0)
RuntimeError: CUDA driver error: out of memory
```

`python test/test_cuda_expandable_segments.py` generates

```
Traceback (most recent call last):
  File "/opt/pytorch/pytorch/test/test_cuda_expandable_segments.py", line 12, in <module>
    exec(compile(open(filepath).read(), filepath, mode='exec'))
  File "/opt/pytorch/pytorch/test/test_cuda.py", line 66, in <module>
    class TestCuda(TestCase):
  File "/opt/pytorch/pytorch/test/test_cuda.py", line 1609, in TestCuda
    @unittest.skipIf(not TEST_CUDNN, 'CUDNN not available')
  File "/usr/local/lib/python3.10/dist-packages/torch/testing/_internal/common_utils.py", line 4628, in wrapped
    self._value = self._cb()
  File "/usr/local/lib/python3.10/dist-packages/torch/testing/_internal/common_cuda.py", line 20, in <lambda>
    TEST_CUDNN = LazyVal(lambda: TEST_CUDA and torch.backends.cudnn.is_acceptable(torch.tensor(1., device=CUDA_DEVICE)))
RuntimeError: handle_0 INTERNAL ASSERT FAILED at "/opt/pytorch/pytorch/c10/cuda/driver_api.cpp":15, please report a bug to PyTorch.
```

This PR intends to fix this issue by adding various dlopen checks to make sure nvml actually exists, and safely fall back to using the older libcuda based features of cuda expandable segments and oom reporting if nvml is not found. 

cc: @ptrblck 

cc @ptrblck @puririshi98

### CODE REVIEW

File: c10/cuda/CUDACachingAllocator.cpp
Comment (eqy) [inline c10/cuda/CUDACachingAllocator.cpp:800]: Would we want to indicate that NVML isn't available in this case e.g., via a warning? I can't tell if the user would be able to see anything in this case.

File: c10/cuda/driver_api.cpp
Comment (eqy) [inline c10/cuda/driver_api.cpp:16]: Is it necessary to have a catch (all) here?

File: c10/cuda/CUDACachingAllocator.cpp
Comment (eqy) [inline c10/cuda/CUDACachingAllocator.cpp:797]: Similar to below, can we avoid a catch-all?

File: c10/cuda/driver_api.cpp
Comment (Fuzzkatt) [reply c10/cuda/driver_api.cpp]: Followed your offline suggestion to change this to a simple if check, thanks.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (Fuzzkatt) [reply c10/cuda/CUDACachingAllocator.cpp]: Followed your offline suggestion to change this to a simple if check, thanks.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (Fuzzkatt) [reply c10/cuda/CUDACachingAllocator.cpp]: I think the context here is that this function will return an empty string if nvml is not available and a string with actual useful information if it is available. This is then used in the out of memory reporting function here: https://github.com/pytorch/pytorch/blob/main/c10/cuda/CUDACachingAllocator.cpp#L1054, https://github.com/pytorch/pytorch/blob/main/c10/cuda/CUDACachingAllocator.cpp#L1126. But from the user's perspective, without nvml they still get a well-formed out of memory error message, just missing the extra information that nvml would report. Do we want to then add some kind of message in this case like "For more detailed out of memory reporting, please install nvml if possible" or something, or would this not matter to most users?

**Comment by albanD** [comment]
@eqy feel free to ping me once this is ready to merge!

**Review by eqy — APPROVED** [review]
LGTM, if there is a well-formed error message I'm good with reducing warning spam.

**Comment by eqy** [comment]
LGTM CC @albanD

File: c10/cuda/CUDACachingAllocator.cpp
Comment (ngimel) [inline c10/cuda/CUDACachingAllocator.cpp:794]: this would be calling `dlopen` on each invocation?

Comment (cyyever) [reply c10/cuda/CUDACachingAllocator.cpp:794]: Need a static once flag.

Comment (eqy) [reply c10/cuda/CUDACachingAllocator.cpp:794]: That's a good catch, could we make this static @Fuzzkatt ?

Comment (Fuzzkatt) [reply c10/cuda/CUDACachingAllocator.cpp]: Good idea, make a fix for this

**Review by ngimel — APPROVED** [review]

File: c10/cuda/CUDACachingAllocator.cpp
Comment (albanD) [inline c10/cuda/CUDACachingAllocator.cpp:795]: Why is the handle stored here if we never use it?

File: c10/cuda/CUDACachingAllocator.cpp
Comment (albanD) [inline c10/cuda/CUDACachingAllocator.cpp:797]: nit: shouldn't this be factored into a `init_driver_api()` function in the driver API file that internally will handle the call_once and can be called from any place where it might need to get loaded. That will avoid potential duplicate dlopen as well as scattered dlopen code.

**Review by albanD — COMMENTED** [review]
Small nits for code organization.

Also any way we could test this by having a subprocess test (see examples in the test suite) that will set the right flags and trigger an oom in the child process and make sure it raises the right error?

File: c10/cuda/CUDACachingAllocator.cpp
Comment (Fuzzkatt) [reply c10/cuda/CUDACachingAllocator.cpp]: Good idea, fixed by moving dlopen into DriverAPI::get_nvml_handle() in newest refactor

**Comment by Fuzzkatt** [comment]
Addressed the code organization nits. Discussed with @albanD offline; to clarify the use case for this PR, this PR was originally for the Jetson setting since nvml is not supported for Jetson in general. On the Jetson setup, the libnvidia-ml.so.1 file doesn't exist, so the dlopen from the original test would fail. In theory, this could also apply to cases in normal devices where nvml is missing from the setup, but this is super unlikely and would pretty much count as a broken environment. The OSS CI doesn't even have Jetson runners so this wouldn't exist naturally there. Thus, there is no need to write a test case for this since it would require mocking around the cuda install on the test runners which is tricky, and the use case is only for Jetson which is not covered by OSS CI anyways.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (albanD) [inline c10/cuda/CUDACachingAllocator.cpp:18]: not needed anymore?

**Review by albanD — APPROVED** [review]
SGTM
Thanks for the update!

File: c10/cuda/CUDACachingAllocator.cpp
Comment (Fuzzkatt) [reply c10/cuda/CUDACachingAllocator.cpp:18]: Good catch, thanks

**Review by albanD — APPROVED** [review]

**Pull request closed without merging**
